rclpy/src/rclpy/names.cpp

```diff
@@ -0,0 +1,289 @@
+// Copyright 2021 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Include pybind11 before rclpy_common/handle.h includes Python.h
+#include <pybind11/pybind11.h>
+
+#include <rcl/context.h>
+#include <rcl/error_handling.h>
+#include <rcl/expand_topic_name.h>
+#include <rcl/rcl.h>
+#include <rcl/remap.h>
+#include <rcl/types.h>
+#include <rcl/validate_topic_name.h>
+#include <rmw/validate_full_topic_name.h>
+#include <rmw/validate_namespace.h>
+#include <rmw/validate_node_name.h>
+
+#include <memory>
+#include <stdexcept>
+#include <string>
+
+#include "rclpy_common/handle.h"
+
+#include "rclpy_common/exceptions.hpp"
+
+#include "context.hpp"
+
+namespace rclpy
+{
+py::object
+get_validation_error_for_topic_name(const char * topic_name)
+{
+  int validation_result;
+  size_t invalid_index;
+  rcl_ret_t ret = rcl_validate_topic_name(topic_name, &validation_result, &invalid_index);
+  if (ret == RCL_RET_BAD_ALLOC) {
+    rcl_reset_error();
+    throw std::bad_alloc();
+  } else if (ret != RCL_RET_OK) {
+    throw RCLError("failed to validate name");
+  }
+
+  if (validation_result == RCL_TOPIC_NAME_VALID) {
+    return py::none();
+  }
+
+  const char * validation_message = rcl_topic_name_validation_result_string(validation_result);
+
+  return py::make_tuple(validation_message, invalid_index);
+}
+
+py::object
+get_validation_error_for_full_topic_name(const char * topic_name)
+{
+  int validation_result;
+  size_t invalid_index;
+  rmw_ret_t ret = rmw_validate_full_topic_name(topic_name, &validation_result, &invalid_index);
+  if (ret == RCL_RET_BAD_ALLOC) {
+    rcl_reset_error();
+    throw std::bad_alloc();
+  } else if (ret != RCL_RET_OK) {
+    throw RCLError("failed to validate name");
+  }
+
+  if (validation_result == RMW_NAMESPACE_VALID) {
+    return py::none();
+  }
+
+  const char * validation_message = rmw_full_topic_name_validation_result_string(validation_result);
+
+  return py::make_tuple(validation_message, invalid_index);
+}
+
+py::object
+get_validation_error_for_namespace(const char * namespace_)
+{
+  int validation_result;
+  size_t invalid_index;
+  rmw_ret_t ret = rmw_validate_namespace(namespace_, &validation_result, &invalid_index);
+  if (ret == RCL_RET_BAD_ALLOC) {
```

Comment:
@sloretz in the same vein as @cottsay comments about `rcutils` error handling vs. `rcl` error handling, it'd be good to deal with `rmw` error handling using `rmw` API.

Reply:
I think there a similar issue in `get_validation_error_for_full_topic_name`, `get_validation_error_for_node_name`.

On further investigation, there seems to be some handling for this in rcl. They seem to be translated to rcl error codes though?
https://github.com/ros2/rcl/blob/0ca8c6162136594e490ba69fbf288de1b5f7073d/rcl/src/rcl/expand_topic_name.c#L79

Reply:
Both `rmw` and `rcl` error handling APIs are aliases of `rcutils`'. That's why people tend to use them interchangeably. Which means we'll have a huge body of code to migrate if that stops being true at some point.

Reply:
Regardless, I think it is worthwhile to use the correct types (even if they are aliases of each other).  Besides the fact that it makes migration possible (though that is unlikely), there is no question for future readers of the code on whether this is correct or not.

Reply:
I added `RMWError` in fc32a0a99ebafc5e28efa343d0e6c868c977d9f7 and used rcl, rcutuils, or rmw constants and functions for each case in c269b6ac92dd3f15aafebffd3117055cfc34cad2